client/rpc_timeout.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2023 CloudWeGo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+
+	"github.com/cloudwego/kitex/client"
+	"github.com/cloudwego/kitex/pkg/klog"
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"github.com/cloudwego/kitex/pkg/rpctimeout"
+	"github.com/kitex-contrib/config-etcd/etcd"
+	"github.com/kitex-contrib/config-etcd/utils"
+)
+
+const defaultTimeoutConfig = "{\"*\": {\"conn_timeout_ms\": 100,\"rpc_timeout_ms\": 2000}}"
+
+// WithRPCTimeout sets the RPC timeout policy from etcd configuration center.
+func WithRPCTimeout(dest, src string, etcdClient etcd.Client, uniqueID int64, opts utils.Options) []client.Option {
+	param, err := etcdClient.ClientConfigParam(&etcd.ConfigParamConfig{
+		Category:          rpcTimeoutConfigName,
+		ServerServiceName: dest,
+		ClientServiceName: src,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	for _, f := range opts.EtcdCustomFunctions {
+		f(&param)
+	}
+	key := param.Prefix + "/" + param.Path
+	return []client.Option{
+		client.WithTimeoutProvider(initRPCTimeoutContainer(key, dest, etcdClient, uniqueID)),
+		client.WithCloseCallbacks(func() error {
+			// cancel the configuration listener when client is closed.
+			etcdClient.DeregisterConfig(key, uniqueID)
+			return nil
+		}),
+	}
+}
+
+func initRPCTimeoutContainer(key, dest string,
+	etcdClient etcd.Client, uniqueID int64,
+) rpcinfo.TimeoutProvider {
+	rpcTimeoutContainer := rpctimeout.NewContainer()
+
+	onChangeCallback := func(isDefault bool, data string, parser etcd.ConfigParser) {
+		if isDefault {
```

Comment:
This will be enough:
```
configs := map...{}

if !restoreDefault {
    ...
}
```

Since `rpcTimeoutContainer.NotifyPolicyChange` will set its inner globalConfig: to be `CopyDefaultRPCTimeout()`